src/messages/mod.rs

```diff
@@ -72,44 +93,53 @@ pub struct HopMessageWithBytes {
 
 impl HopMessageWithBytes {
     /// Serialize message and keep both SignedRoutingMessage and Bytes.
-    pub fn new(content: SignedRoutingMessage, log_ident: &LogIdent) -> Result<Self> {
-        let message = Message::Hop(content);
+    pub fn new(full_content: SignedRoutingMessage, log_ident: &LogIdent) -> Result<Self> {
+        let message = Message::Hop(full_content);
         let full_message_bytes = to_network_bytes(&message)?;
 
-        let content = if let Message::Hop(content) = message {
-            content
+        let full_content = if let Message::Hop(full_content) = message {
+            full_content
         } else {
             unreachable!("Created as Hop can only match Hop.")
         };
 
-        let message = Self::new_from_parts(content, full_message_bytes);
+        let peek_msg = PeekedSignedRoutingMessageInfo {
+            dst: full_content.routing_message().dst,
+        };
+        let message = Self::new_from_parts(Some(full_content), peek_msg, full_message_bytes);
 
         trace!(
             "{} Creating message hash({:?}) {:?}",
             log_ident,
             message.full_message_crypto_hash,
-            message.content.routing_message(),
+            message
+                .full_content
+                .as_ref()
+                .expect("New HopMessageWithBytes need full_content")
+                .routing_message(),
         );
 
         Ok(message)
     }
 
-    fn new_from_parts(content: SignedRoutingMessage, full_message_bytes: Bytes) -> Self {
+    fn new_from_parts(
+        full_content: Option<SignedRoutingMessage>,
+        peek_msg: PeekedSignedRoutingMessageInfo,
+        full_message_bytes: Bytes,
+    ) -> Self {
         let full_message_crypto_hash = crypto::sha3_256(&full_message_bytes);
 
         Self {
-            content,
+            full_content,
+            peek_msg,
             full_message_bytes,
             full_message_crypto_hash,
         }
     }
 
-    pub fn signed_routing_message(&self) -> &SignedRoutingMessage {
-        &self.content
-    }
-
-    pub fn into_signed_routing_message(self) -> SignedRoutingMessage {
-        self.content
+    pub fn take_or_deserialize_signed_routing_message(&mut self) -> Result<SignedRoutingMessage> {
```

Comment:
Is such verbose name necessary? Isn't the source of the message an implementation detail that the caller doesn't need to be concerned with? I.e., just call it `take_signed_routing_message`.

Reply:
take imply a quick move, I would want to make sure it is clear what is happening.
Since it is used so few times is that an issue?

Reply:
Then maybe `extract_signed_routing_message`. Or leave it as is if you prefer. Not a big issue for me.